certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,131 @@
+    require threshold > 0;
+    require guardianStorageContract.entries[safeContract].count < max_uint256; // To prevent overflow (Realistically can't reach).
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+    require guardian != otherGuardian;
+
+    require !safeContract.isOwner(guardian);
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    require threshold < guardianStorageContract.countGuardians(safeContract);
+
+    bool otherWasGuardian = currentContract.isGuardian(safeContract, otherGuardian);
+
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted => currentContract.isGuardian(safeContract, guardian);
+    assert otherWasGuardian == currentContract.isGuardian(safeContract, otherGuardian);
+}
+
+// This integrity rule verifies that the revokeGuardianWithThreshold(...) works as expected.
+rule removeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold) {
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+
+    // This could be removed once linked list invariant is implemented.
+    // <------------------------------------------------------------>
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+    require guardian != prevGuardian;
+    require guardian != nextGuardian;
+    require prevGuardian != nextGuardian;
+    // <------------------------------------------------------------>
+
+    currentContract.revokeGuardianWithThreshold(e, safeContract, prevGuardian, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert !currentContract.isGuardian(safeContract, guardian);
+    assert guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian;
+}
+
+// This integrity rule verifies that the guardian can always be removed considering ideal conditions.
+rule guardianCanAlwaysBeRemoved(env e, address guardian, address prevGuardian, uint256 threshold) {
+
+    requireSocialRecoveryModuleEnabled();
+
+    require e.msg.value == 0;
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
```

Comment:
This can also be replaced by the same as :point_up: right?

Reply:
Based on the count? Or something different?

This case ensures that if the count is marked (from the initial state) as more than 0, then there should be some guardian which starts from the `SENTINAL`. And we can ensure that the addition happens in that case (from the `SENTINAL`).

Reply:
Done 👍🏾